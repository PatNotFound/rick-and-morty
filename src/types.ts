type UrlFilterType = {
  name: string;
  status: string;
  species: string;
  gender: string;
};

type CharacterType = {
  created: string;
  episode: string[];
  gender: string;
  id: number;
  image: string;
  location: { name: string; url: string };
  name: string;
  origin: { name: string; url: string };
  species: string;
  status: string;
  type: string;
  url: string;
};

type ModalProps = {
  toggleVisibility: () => void;
  data: { label: string; value: string | number }[];
  title: string;
  image: string;
};

type FilterType = {
  name: string;
  status: string;
  species: string;
  gender: string;
};

type FilterPropsType = {
  onFilterChange: (e: URLSearchParams) => void;
  searchParams: URLSearchParams;
};

export type {
  UrlFilterType,
  CharacterType,
  ModalProps,
  FilterPropsType,
  FilterType,
};
