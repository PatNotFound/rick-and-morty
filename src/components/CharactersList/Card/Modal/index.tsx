import React from 'react';
import { useEffect } from 'react';
import { createPortal } from 'react-dom';

export type ModalProps = {
  toggleVisibility: () => void;
  data: { label: string; value: string | number }[];
  title: string;
  image: string;
};

const Modal = ({
  toggleVisibility,
  title,
  image,
  data,
}: Readonly<ModalProps>): JSX.Element | null => {
  useEffect(() => {
    const onEscape = (e: KeyboardEvent) => {
      if (e.key === 'Escape') toggleVisibility();
    };

    window.addEventListener('keydown', onEscape);
    return () => window.removeEventListener('keydown', onEscape);
  }, []);

  const modal: JSX.Element = (
    <>
      <div
        className="fixed inset-0 bg-gray-400 bg-opacity-60"
        onClick={toggleVisibility}
        aria-label={`${title} modal`}
      />
      <div
        className="fixed left-2/4 top-2/4 flex w-auto min-w-[500px] -translate-x-1/2 -translate-y-1/2 transform flex-col rounded-xl bg-white p-6"
        aria-modal
        aria-label="Modal Details"
        role="dialog"
        onClick={(e) => e.stopPropagation()}
      >
        <button
          className="absolute right-6 appearance-none self-baseline text-xl outline-none"
          aria-label="Close Modal"
          onClick={toggleVisibility}
          type="button"
        >
          <i className="fa-solid fa-xmark"></i>
        </button>
        <div className="flex flex-row">
          <div className="mr-6 flex min-w-[150px] flex-col items-center justify-center">
            <img
              src={image}
              alt={title}
              className="mb-4 h-24 rounded-full border border-gray-50 shadow-xl"
            />
            <header className="flex text-center text-xl font-bold text-green-500">
              {title}
            </header>
          </div>
          <ul className="flex w-full flex-col divide-y">
            {data.map(({ label, value }) => (
              <li className="mt-2 flex flex-col" key={label}>
                <label className="mr-1 text-lg font-medium">{label}</label>
                <span>{value}</span>
              </li>
            ))}
          </ul>
        </div>
      </div>
    </>
  );

  return createPortal(modal, document.body);
};

export default Modal;
