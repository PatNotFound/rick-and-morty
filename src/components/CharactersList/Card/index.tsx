import React from 'react';
import { ReactElement, useState } from 'react';
import { CharacterType } from '../../../types/character';
import Modal from './Modal';

const COLOR = {
  Alive: 'text-green-600',
  Dead: 'text-red-600',
  unknown: 'text-gray-600',
};

function Card(character: CharacterType): ReactElement {
  const {
    episode,
    gender,
    id,
    image,
    location: { name: locationName },
    name,
    origin: { name: originName },
    species,
    status,
  } = character;

  const [modalIsVisible, setModalIsVisible] = useState(false);
  const toggleModalVisibility = () => setModalIsVisible(!modalIsVisible);

  return (
    <li>
      <button
        className="flex w-full cursor-pointer items-center gap-4 rounded-xl p-6 shadow-xl"
        onClick={() => toggleModalVisibility()}
        aria-label={`${name} card`}
      >
        <img src={image} alt={`${name}-${id}`} className="h-24 rounded-full" />
        <div className="truncate text-left">
          <div className="truncate text-lg font-medium">{name}</div>
          <div className="flex items-center gap-1">
            <i
              className={`fa-solid fa-circle text-xs ${
                COLOR[status as keyof typeof COLOR]
              } opacity-40`}
            />
            <span className="capitalize">{status}</span>
          </div>
          <div className="flex">
            {species === 'Human' ? 'ðŸ‘¤' : 'ðŸ‘½'}{' '}
            <div className="ml-1 truncate capitalize">{species}</div>
          </div>
        </div>
        {modalIsVisible && (
          <Modal
            toggleVisibility={toggleModalVisibility}
            title={name}
            image={image}
            data={[
              { label: 'Status', value: status },
              { label: 'Species', value: species },
              { label: 'Gender', value: gender },
              { label: 'Location', value: locationName },
              { label: 'Origin', value: originName },
              { label: 'Number of appearances', value: episode.length },
            ]}
          />
        )}
      </button>
    </li>
  );
}

export default Card;
