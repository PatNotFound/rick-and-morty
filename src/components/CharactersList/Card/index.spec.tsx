import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Card from ".";
import { CharacterType } from "../../../types/character";

describe("Card", () => {
  const getRender = (props: CharacterType) => render(<Card {...props} />);

  const mockedProps = {
    created: "2017-11-04T18:50:21.651Z",
    episode: [
      "https://rickandmortyapi.com/api/episode/1",
      "https://rickandmortyapi.com/api/episode/2",
    ],
    gender: "Male",
    id: 2,
    image: "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
    location: {
      name: "Citadel of Ricks",
      url: "https://rickandmortyapi.com/api/location/3",
    },
    name: "Morty Smith",
    origin: { name: "unknown", url: "" },
    species: "Human",
    status: "Alive",
    type: "",
    url: "https://rickandmortyapi.com/api/character/2",
  };

  it("should render the component", () => {
    const { container } = getRender(mockedProps);

    expect(container).toMatchSnapshot();
  });

  it("should trigger the modal when clicking the card", () => {
    const { getByLabelText } = getRender(mockedProps);

    const card = getByLabelText(`${mockedProps.name}-card`);

    fireEvent.click(card);

    const modal = getByLabelText(`${mockedProps.name}-modal`);

    expect(modal).toBeInTheDocument();
  });
});
