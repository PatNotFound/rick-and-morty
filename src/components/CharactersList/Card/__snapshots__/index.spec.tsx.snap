// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render the component 1`] = `
<div>
  <li>
    <button
      aria-label="Morty Smith-card"
      class="flex w-full cursor-pointer items-center gap-4 rounded-xl p-6 shadow-xl"
    >
      <img
        alt="Morty Smith-2"
        class="h-24 rounded-full"
        src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      />
      <div
        class="truncate text-left"
      >
        <div
          class="truncate text-lg font-medium"
        >
          Morty Smith
        </div>
        <div
          class="flex items-center gap-1"
        >
          <i
            class="fa-solid fa-circle text-xs text-green-600 opacity-40"
          />
          <span
            class="capitalize"
          >
            Alive
          </span>
        </div>
        <div
          class="flex"
        >
          ðŸ‘¤
           
          <div
            class="ml-1 truncate capitalize"
          >
            Human
          </div>
        </div>
      </div>
    </button>
  </li>
</div>
`;
